### PR TITLE
hwcontext_qsv: change qsv download/upload to child download/upload

### PR DESCRIPTION
**Pull request by tong1wu**
qsv is using MSDK VPP API to process download/upload, now change it to
using child context download/upload methods.

This change can fix the issue on Windows of cm_copy returning wrong
frames.

### CODE REVIEW

**Comment by tong1wu** [comment]
@xhaihao pls help to review the patch set, thanks

File: libavutil/hwcontext_dxva2.c [hunk 1]
```diff
@@ -316,10 +316,10 @@ static int dxva2_map_frame(AVHWFramesContext *ctx, AVFrame *dst, const AVFrame *
         goto fail;
     }
 
-    for (i = 0; i < nb_planes; i++)
+    for (i = 0; i < 4; i++)
```
Comment (xhaihao) [inline libavutil/hwcontext_dxva2.c:319]: Why is this change needed ?

**Comment by xhaihao** [comment]
I can't use hwupload to load nv12 data to gfx memory for dxva2, does your patch work when the child device type for qsv is dxva2 ? 

$ ffmpeg.exe -init_hw_device dxva2=dv -f lavfi -i yuvtestsrc -vf "format=nv12,hwupload" -f null -
[...]
[Parsed_hwupload_1 @ 000001AC14243380] Query format failed for 'Parsed_hwupload_1': Invalid argument
Error reinitializing filters!
Failed to inject frame into filter network: Invalid argument
Error while processing the decoded data for stream #0:0
Conversion failed!

**Comment by tong1wu** [comment]
> I can't use hwupload to load nv12 data to gfx memory for dxva2, does your patch work when the child device type for qsv is dxva2 ?
> 
> $ ffmpeg.exe -init_hw_device dxva2=dv -f lavfi -i yuvtestsrc -vf "format=nv12,hwupload" -f null - [...] [Parsed_hwupload_1 @ 000001AC14243380] Query format failed for 'Parsed_hwupload_1': Invalid argument Error reinitializing filters! Failed to inject frame into filter network: Invalid argument Error while processing the decoded data for stream #0:0 Conversion failed!



> I can't use hwupload to load nv12 data to gfx memory for dxva2, does your patch work when the child device type for qsv is dxva2 ?
> 
> $ ffmpeg.exe -init_hw_device dxva2=dv -f lavfi -i yuvtestsrc -vf "format=nv12,hwupload" -f null - [...] [Parsed_hwupload_1 @ 000001AC14243380] Query format failed for 'Parsed_hwupload_1': Invalid argument Error reinitializing filters! Failed to inject frame into filter network: Invalid argument Error while processing the decoded data for stream #0:0 Conversion failed!

Hi haihao, I checked this error. It's not caused by this patch. It's because hwcontext_dxva2 doesn't have get_constraints function. https://github.com/FFmpeg/FFmpeg/blob/master/libavutil/hwcontext.c#L584 returns and reports error. This error has always existed and maybe no one ever use dxva2 upload. I may submit another pr to add the get_constraints function to enable this path. The patch here fixes the ffmpeg-libmfx:dxva2 uploading.

**Comment by tong1wu** [comment]
> I can't use hwupload to load nv12 data to gfx memory for dxva2, does your patch work when the child device type for qsv is dxva2 ?
> 
> $ ffmpeg.exe -init_hw_device dxva2=dv -f lavfi -i yuvtestsrc -vf "format=nv12,hwupload" -f null - [...] [Parsed_hwupload_1 @ 000001AC14243380] Query format failed for 'Parsed_hwupload_1': Invalid argument Error reinitializing filters! Failed to inject frame into filter network: Invalid argument Error while processing the decoded data for stream #0:0 Conversion failed!



> I can't use hwupload to load nv12 data to gfx memory for dxva2, does your patch work when the child device type for qsv is dxva2 ?
> 
> $ ffmpeg.exe -init_hw_device dxva2=dv -f lavfi -i yuvtestsrc -vf "format=nv12,hwupload" -f null - [...] [Parsed_hwupload_1 @ 000001AC14243380] Query format failed for 'Parsed_hwupload_1': Invalid argument Error reinitializing filters! Failed to inject frame into filter network: Invalid argument Error while processing the decoded data for stream #0:0 Conversion failed!

Hi haihao, I checked this error. It's not caused by this patch. It's because hwcontext_dxva2 doesn't have get_constraints function. https://github.com/FFmpeg/FFmpeg/blob/master/libavutil/hwcontext.c#L584 returns and reports error. This error has always existed and maybe no one ever use dxva2 upload. I may submit another pr to add the get_constraints function to enable this path. The patch here fixes the ffmpeg-libmfx:dxva2 uploading.

File: libavutil/hwcontext_dxva2.c (hunk 1 above)
Comment (tong1wu) [reply libavutil/hwcontext_dxva2.c:319]: because according to https://github.com/FFmpeg/FFmpeg/blob/master/libavutil/hwcontext_dxva2.c#L287, dst->format is hwframe when uploading and nb_planes is returned as 0. https://github.com/FFmpeg/FFmpeg/blob/master/libavutil/hwcontext_d3d11va.c#L374 It's also 4 instead of nb_planes in d3d11. 
If this changed accepted, I will update it to delete the nb_planes variable since it is only used here. Thanks.

**Comment by tong1wu** [comment]
@xhaihao hi haihao, is there any other concern for this pr? thanks.

**Comment by tong1wu** [comment]
Going to close it. This issue is due to cmcopy. It's VPL's responsibility to fix it.
